R/preprocess.R

```diff
@@ -21,41 +21,127 @@ enw_metadata <- function(obs, target_date = "reference_date") {
+#'  * `week`, numeric, 0 based from start of time series
+#'  * `month`, numeric, 0 based from start of time series
+#'
 #' @family preprocess
 #' @export
 #' @importFrom data.table as.data.table
-enw_add_metaobs_features <- function(metaobs, holidays = c(),
-                                     holidays_to = "Sunday") {
-  # add days of week
-  metaobs <- data.table::copy(metaobs)
-  metaobs[, day_of_week := weekdays(date)]
-
-  # make holidays be Sundays
-  if (length(holidays) != 0) {
-    metaobs[get(holidays) == TRUE, day_of_week := holidays_to]
+#' @importFrom lubridate week month
+#' @examples
+#'
+#' # make some example date
+#' nat_germany_hosp <- subset(
+#'   epinowcast::germany_covid19_hosp,
+#'   location == "DE" & age_group == "80+"
+#' )[1:40]
+#'
+#' basemeta <- epinowcast::enw_add_metaobs_features(
+#'   nat_germany_hosp,
+#'   datecol = "report_date"
+#' )
+#' basemeta
+#'
+#' # with holidays - n.b.: holidays not found are silently ignored
+#' holidaymeta <- epinowcast::enw_add_metaobs_features(
+#'   nat_germany_hosp,
+#'   datecol = "report_date",
+#'   holidays_to = "Holiday",
+#'   holidays = c(
+#'     "2021-04-04", "2021-04-05",
+#'     "2021-05-01", "2021-05-13",
+#'     "2021-05-24"
+#'   )
+#' )
+#' holidaymeta
+#' subset(holidaymeta, day_of_week == "Holiday")
+enw_add_metaobs_features <- function(
+  metaobs,
+  holidays_to = "Sunday",
+  holidays,
+  datecol = "date"
+) {
+  metaobs <- data.table::as.data.table(metaobs)
+  if (is.null(metaobs[[datecol]])) {
+    stop(sprintf("metaobs does not have datecol '%s'.", datecol))
+  } else if (!is.Date(metaobs[[datecol]])) {
+    stop(sprintf("metaobs column '%s' is not a Date.", datecol))
   }
 
-  # make day of week a factor
-  metaobs[, day_of_week := factor(day_of_week)]
+  # if not present, add day_of_week
+  if (is.null(metaobs$day_of_week)) {
+    metaobs[, day_of_week := weekdays(get(datecol))]
+  }
 
-  # add day feature
-  metaobs[, day := as.numeric(date)]
-  metaobs[, day := day - min(day)]
+  # set holidays to associated day_of_week
+  if (!missing(holidays) && (length(holidays) > 0)) {
+    holidays <- as.IDate(holidays, optional = TRUE)
+    if (any(is.na(holidays))) {
+      stop(
+        sprintf(
+          "Failed to parse holidays; first failure at index %i.",
+          which.max(is.na(holidays))
+        )
+      )
+    }
+    metaobs[get(datecol) %in% holidays, day_of_week := holidays_to]
+  }
+
+  # make day_of_week a factor
+  metaobs[, day_of_week := factor(day_of_week)]
 
-  # add week feature
-  metaobs[, week := lubridate::week(date)]
-  metaobs[, week := week - min(week)]
+  # function for transforming numbers to be referenced from 0
+  zerobase <- function(x) {
+    return(x - min(x))
+  }
 
-  # add month feature
-  metaobs[, month := lubridate::month(date)]
-  metaobs[, month := month - min(month)]
+  # only compute columns not already present
+  tarcols <- setdiff(c("day", "week", "month"), colnames(metaobs))
```

Comment:
I think we need to make this recompute but with a warning as this needs to be called multiple times annoyingly (when we compute Rt we need to extend one of these data. frames in model setup)